### PR TITLE
8309717: C2: Remove Arena::move_contents usage

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

Instead of using `Arena::move_contents` we can just see the arena swap as a form of double buffering, reducing this to a pointer swap and a clear. This allows us to remove `Arena::move_contents`, cleaning up the arena code.

Since this requires allocating another pointer for `Compile`, I took the time to move some members around in order to reduce the padding. This means that this patch does *not* introduce a size change for `Compile`.

I'm currently running tier1-3 tests.

Thanks for considering this,
Johan

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309717](https://bugs.openjdk.org/browse/JDK-8309717): C2: Remove Arena::move_contents usage (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14391/head:pull/14391` \
`$ git checkout pull/14391`

Update a local copy of the PR: \
`$ git checkout pull/14391` \
`$ git pull https://git.openjdk.org/jdk.git pull/14391/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14391`

View PR using the GUI difftool: \
`$ git pr show -t 14391`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14391.diff">https://git.openjdk.org/jdk/pull/14391.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14391#issuecomment-1584349766)

### CODE REVIEW

**Comment by jdksjolen** [comment]
Passes the tests.

File: src/hotspot/share/opto/compile.hpp [hunk 1]
```diff
@@ -795,8 +808,8 @@ class Compile : public Phase {
   uint         unique() const              { return _unique; }
   uint         next_unique()               { return _unique++; }
   void         set_unique(uint i)          { _unique = i; }
-  Arena*       node_arena()                { return &_node_arena; }
-  Arena*       old_arena()                 { return &_old_arena; }
+  Arena*       node_arena()                { return _node_arena; }
+  Arena*       old_arena()                 { return &_node_arena_one == _node_arena ? &_node_arena_two : &_node_arena_one; }
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.hpp:812]:
```suggestion
  Arena*       old_arena()                 { return (&_node_arena_one == _node_arena) ? &_node_arena_two : &_node_arena_one; }
```

File: src/hotspot/share/opto/matcher.cpp [hunk 2]
```diff
@@ -335,7 +335,7 @@ void Matcher::match( ) {
   Node* new_ideal_null = ConNode::make(TypePtr::NULL_PTR);
 
   // Swap out to old-space; emptying new-space
-  Arena *old = C->node_arena()->move_contents(C->old_arena());
+  Arena *old = C->swap_old_and_new();
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/matcher.cpp:338]:
```suggestion
  Arena* old = C->swap_old_and_new();
```

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me.

**Comment by jdksjolen** [comment]
Thanks Tobias, I took your changes and applied them.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]

**Comment by jdksjolen** [comment]
Thank you for the reviews.

/integrate

**Pull request closed without merging**